kafka/socks5_wrapper.py

```diff
@@ -0,0 +1,236 @@
+                    )
+                    self._state = ProxyConnectionStates.AUTHENTICATING
+
+        if self._state == ProxyConnectionStates.AUTHENTICATING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:2] == b"\x01\x00":
+                    self._state = ProxyConnectionStates.REQUEST_SUBMIT
+                else:
+                    log.error("Socks5 proxy authentication failure")
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+        if self._state == ProxyConnectionStates.REQUEST_SUBMIT:
+            if self._target_afi == socket.AF_INET:
+                addr_type = 1
+                addr_len = 4
+            elif self._target_afi == socket.AF_INET6:
+                addr_type = 4
+                addr_len = 16
+            else:
+                log.error("Unknown address family, %r", self._target_afi)
+                self._state = ProxyConnectionStates.DISCONNECTED
+                self._sock.close()
+                return errno.ECONNREFUSED
+            self._buffer_out = struct.pack(
+                "!bbbb{}sh".format(addr_len),
+                5,  # version
+                1,  # command: connect
+                0,  # reserved
+                addr_type,
+                socket.inet_pton(self._target_afi, addr[0]),
+                addr[1],
+            )
+            self._state = ProxyConnectionStates.REQUESTING
+
+        if self._state == ProxyConnectionStates.REQUESTING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:2] == b"\x05\x00":
+                    self._state = ProxyConnectionStates.READ_ADDRESS
+                else:
+                    log.error("Proxy request failed: %r", buf[1:2])
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+        if self._state == ProxyConnectionStates.READ_ADDRESS:
+            # we don't really care about the remote endpoint address, but need to clear the stream
+            buf = self._peek_buf(2)
+            if buf:
+                if buf[0:2] == b"\x00\x01":
+                    _ = self._read_buf(2 + 4 + 2)
+                elif buf[0:2] == b"\x00\x05":
+                    _ = self._read_buf(2 + 16 + 2)
+                else:
+                    log.error("Unrecognized remote address type %r", buf[1:2])
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+                self._state = ProxyConnectionStates.COMPLETE
+
+        if self._state == ProxyConnectionStates.COMPLETE:
+            return 0
+
+        # not reached
+        log.error("Internal error, state %r not handled correctly", self._state)
```

Comment:
Currently this place might be reached in several cases and determining the exact condition that triggered it may not be possible. I think explicit else branches for all possibly failing cases earlier in the logic would be better (possibly with some helper function that allows repetitive code, the same error handling is already repeated quite a few time).

Reply:
This condition should never be reached; I was considering between raising an exception or just returning general connection error. The log output does include the current state, which should help to debug somewhat on why things fell through.

In actual operation, we also have the remote proxy logs available for decoding what could have been going wrong.

Reply:
It can currently be reached at least if some reads are short (we get 0-1 bytes when expecting two) or if response to NEGOTIATE_PROPOSE request is unexpected.

Reply:
Sorry not when reads are short but when output buffer cannot be flushed.

Reply:
I think I need to add a bit of documentation on this. This code is called with sockets in non-blocking mode, so failing to flush / send all of buffer would result in raising EAGAIN/EWOULDBLOCK. This code specifically avoids catching those and let the retry logic in kafka/conn.py to call back to connect_ex either on timer or using select() & friends.

Reply:
Ah, OK. Then the entire checking return value part is a bit confusing if the call is never expected to return a failure but raise an error instead (plus that NEGOTIATE_PROPOSE still looks to be missing else branch for the case where response isn't one of the two we expect).